metasv/generate_final_vcf.py

```diff
@@ -154,6 +155,115 @@ def filter_confused_INS_calls(nonfilterd_bed, filterd_bed, wiggle=20):
+
+def filter_itxs(feature):
+    n=len(feature.fields)/2
+    del_interval_idp=map(int,feature.fields[7].split("-"))
+    del_interval_itx_1=map(int,feature.fields[7].split(",")[0].split("-"))
+    del_interval_itx_2=map(int,feature.fields[7].split(",")[1].split("-"))
+    if filter(lambda x:abs(x[0]-del_interval_idp[0])+abs(x[1]-del_interval_idp[1]) == 0 ,
+                       [del_interval_itx_1,del_interval_itx_2]):
+        return None
+    return pybedtools.Interval(feature.chrom, feature.start, feature.end, name=feature.name, 
+                               otherfields=feature.fields[6:n])        
+
+def merge_idp_itx(record_dup,records_del,del_pos,del_interval,svtype):
+    record_dup.ALT = [vcf.model._SV("IDP")]
+    record_dup.INFO["%s_POS"%svtype]=int(del_pos)
+    record_dup.INFO["%s_INTERVAL"%svtype]=del_interval
+    record_dup.INFO["SVTYPE"]=svtype
+    record_dup.INFO["SVMETHOD"]=list(set(reduce(lambda y,z:y+z,map(lambda x: x.INFO["SVMETHOD"]
+                                                                   ,[record_dup]+records_del))))
+    record_dup.INFO["SOURCES"]=",".join(map(lambda x: x.INFO["SOURCES"],[record_dup]+records_del))
+    record_dup.INFO["NUM_SVMETHODS"]=len(record_dup.INFO["SVMETHOD"])
+    record_dup.INFO["NUM_SVTOOLS"]=len(set(map(lambda x: x.split('-')[-1],record_dup.INFO["SOURCES"].split(','))))
+    return record_dup
+
+def resolve_for_IDP_ITX(vcf_records,pad=0,wiggle=10):
+    del_records = filter(lambda x: (x.INFO["SVTYPE"] == "DEL") and (x.FILTER != "LowQual"),vcf_records)
+    dup_records = filter(lambda x: (x.INFO["SVTYPE"] == "DUP") and (x.FILTER != "LowQual"),vcf_records)
+    other_records = filter(lambda x: (x.INFO["SVTYPE"] not in ["DEL","DUP"]) or x.FILTER == "LowQual",vcf_records)
+    del_bedtool = pybedtools.BedTool([pybedtools.Interval(x.CHROM, x.POS, (x.start+abs(x.INFO["SVLEN"])),
+                                      name="DEL_%d"%i) for i,x in enumerate(del_records)])     
+    dup_bedtool = pybedtools.BedTool([pybedtools.Interval(x.CHROM, x.POS, (x.start+abs(x.INFO["SVLEN"])),
+                                      name="DUP_%d"%i) for i,x in enumerate(dup_records)])     
+    idp_bedtool=dup_bedtool.window(del_bedtool,w=wiggle).each(partial(find_idp,wiggle=wiggle)).sort()
+    remained_dup_bedtool=dup_bedtool.subtract(idp_bedtool,A=True,f=0.95,r=True).sort()
+    remained_del_bedtool=del_bedtool.subtract(idp_bedtool.each(partial(extract_del_interval)).sort(),A=True,f=0.95,r=True)
+    itx_bedtool=idp_bedtool.window(idp_bedtool,w=wiggle).each(partial(find_itx,wiggle=wiggle)).sort()
+    remained_idp_bedtool_1=idp_bedtool.window(itx_bedtool,w=wiggle).each(partial(filter_itxs)).sort() 
+    remained_idp_bedtool_2=idp_bedtool.window(itx_bedtool,w=wiggle,c=True).filter(lambda x:x.fields[-1]=="0").cut(range(idp_bedtool.field_count())).sort()
+    vcf_records = other_records + [dup_records[int(x.name.split("_")[-1])] for x in remained_dup_bedtool] + \
```

Comment:
We definitely need to use something cleaner here and at other places. Perhaps, extending the Interval class from pybedtools might be a better way. Something we should keep in mind for a cleanup.